_posts/2021-11-18-keras-updates/keras-updates.Rmd

```diff
@@ -0,0 +1,159 @@
+
+In R, between present-from-the-outset packages `tensorflow` and `keras`, responsibilities have always been distributed the way they are now: `tensorflow` providing indispensable basics, but often, remaining completely transparent to the user; `keras` being the thing you use in your code. In fact, it is possible to train a Keras model without ever consciously using `tensorflow`.
+
+On the Python side, things have been undergoing significant changes, ones where, in some sense, the latter development has been inverting the first. In the beginning, TensorFlow and [Keras](keras.io) were separate libraries, with TensorFlow providing a backend -- one among several -- for Keras to make use of. At some point, Keras code got incorporated into the TensorFlow codebase. Finally (as of today), following an extended period of slight confusion, Keras got moved out again, and has started to -- again -- considerably grow in features.
+
+It is just that quick growth that has created, on the R side, the need for extensive low-level refactoring and enhancements. (Of course, the user-facing new functionality itself also had to be implemented!)
+
+Before we get to the promised highlights, a word on how we think about Keras.
+
+## Have your cake and eat it, too: A philosophy of (R) Keras
+
+If you've used Keras in the past, you know what it's always been intended to be: a high-level library, making it easy (as far as such a thing *can* be easy) to train neural networks in R. Actually, it's not just about *easy*. Keras would like to enable users to write natural-feeling, idiomatic-looking code. This, to a high degree, is achieved by its allowing for object composition though the pipe operator; it is also a consequence of its abundant wrappers, convenience functions, and functional (stateless) semantics.[^3]
+
+[^3]: Provided stateless semantics is possible, in some context. This is anything but a matter of course, with an object-oriented language like Python in the background.
+
+However, due to the way TensorFlow and Keras have developed on the Python side -- referring to the big architectural and semantic changes between versions 1.x and 2.x, first comprehensively characterized on this blog [here](https://blogs.rstudio.com/ai/posts/2019-10-08-tf2-whatchanges/) -- it has become more challenging to provide, to the R user, all of the functionality available on the Python side. In addition, maintaining compatibility with several versions of Python TensorFlow -- something R Keras has always been doing -- by necessity gets more and more challenging, the more wrappers and convenience functions you add.
```

Comment:
```suggestion
However, due to the way TensorFlow and Keras have developed on the Python side -- referring to the big architectural and semantic changes between versions 1.x and 2.x, first comprehensively characterized on this blog [here](https://blogs.rstudio.com/ai/posts/2019-10-08-tf2-whatchanges/) -- it has become more challenging to provide all of the functionality available on the Python side to the R user. In addition, maintaining compatibility with several versions of Python TensorFlow -- something R Keras has always done -- by necessity gets more and more challenging, the more wrappers and convenience functions you add.
```

Reply:
Since that first sentence is structurally complex, I moved that clause so it reads "in-line."

Reply:
oh yeah, much better!